### DISCUSSION TITLE
EVP_Digest*() routines return a value

### DISCUSSION THREAD

**Discussion by JohnnySavages** [openssl/openssl#23420] [Q&A] [answered]
There was an old commit that added a return code for EVP_Digest*(): https://github.com/openssl/openssl/commit/2dc769a1c17e1e0c7aef6e11496c8ba2c1db2e28. But all `<mech>_Init`, `<mech>_Update`, `<mech>_Final` functions always return 1.

My question is: why do we need return codes nowadays, are they still relevant? Maybe we should use them as void functions?

**Comment by t8m** [comment, marked as answer]
> But all `<mech>_Init`, `<mech>_Update`, `<mech>_Final` functions always return 1.

That is absolutely not true. They can definitely return 0 for failures. Are you looking at the current source code of OpenSSL?

**Reply by JohnnySavages** [reply]
https://github.com/openssl/openssl/blob/0f644b96d209443b4566f7e86e3be2568292e75b/crypto/sha/sha256.c#L27, https://github.com/openssl/openssl/blob/0f644b96d209443b4566f7e86e3be2568292e75b/include/crypto/md32_common.h#L154, https://github.com/openssl/openssl/blob/0f644b96d209443b4566f7e86e3be2568292e75b/include/crypto/md32_common.h#L217.
Perhaps I missed some specific mechanism.

**Reply by t8m** [reply]
Ah, you were referring to the low-level algorithm _Init _Update _Final functions. Yeah, these return just 1. But that is no problem. We cannot change their signatures as these functions are public although deprecated. Once they are no longer public we could change their signatures but that is something for future.

Also please note that EVP_DigestInit and other EVP high level functions can definitely return 0 in some circumstances regardless of the infallibility of the low-level functions.
